### PR TITLE
Request for VACUUM after deleting GPKG layer

### PR DESCRIPTION
**Pull request by signedav**
After deleting a layer of a GeoPackage it requests user if it should VACUUM the package.

image

### CODE REVIEW

**Comment by elpaso** [comment]
I like the idea but I don't like having two modal dialogs after a deletion, can we skip the initial information dialog?

**Comment by signedav** [comment]
I completely agree @elpaso I'll put the original message into else...

File: src/providers/ogr/qgsgeopackagedataitems.cpp [hunk 1]
```diff
@@ -525,9 +525,23 @@ bool QgsGeoPackageAbstractLayerItem::deleteLayer()
   }
   else
   {
-    QMessageBox::information( nullptr, tr( "Delete Layer" ), tr( "Layer <b>%1</b> deleted successfully." ).arg( mName ) );
     if ( mParent )
+    {
+      QgsGeoPackageConnectionItem *connectionParentItem = qobject_cast<QgsGeoPackageConnectionItem *>( mParent );
+      if ( connectionParentItem )
+      {
+        if ( QMessageBox::question( nullptr, QObject::tr( "Delete Layer" ), QObject::tr( "The layer <b>%1</b> deleted successfully."
```
Comment (elpaso) [inline src/providers/ogr/qgsgeopackagedataitems.cpp]: My english is bad but maybe better:
```suggestion
        if ( QMessageBox::question( nullptr, QObject::tr( "Delete Layer" ), QObject::tr( "The layer <b>%1</b> was deleted successfully."
```

Comment (m-kuhn) [inline src/providers/ogr/qgsgeopackagedataitems.cpp]: Should this be equal to the other message?

```suggestion
        if ( QMessageBox::question( nullptr, QObject::tr( "Delete Layer" ), QObject::tr( "Layer <b>%1</b> deleted successfully."
```

Comment (signedav) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: Hm... It's for sure better. But probably "has been deleted"? Not sure.

Comment (m-kuhn) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: I would also think "has been" deleted if I go by my feeling, but better get confirmation from @nyalldawson or @NathanW2 or @timlinux or someone else who is native

Comment (m-kuhn) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: https://english.stackexchange.com/a/101351

**Review by m-kuhn — APPROVED** [review]

File: src/providers/ogr/qgsgeopackagedataitems.cpp [hunk 2]
```diff
@@ -525,9 +525,23 @@ bool QgsGeoPackageAbstractLayerItem::deleteLayer()
...
+    {
+      if ( QMessageBox::question( nullptr, QObject::tr( "Delete Layer" ), QObject::tr( "The layer <b>%1</b> has been deleted successfully."
+                                  " Compact database (VACUUM) <b>%2</b> now?" ).arg( mName, connectionParentItem->name() ), QMessageBox::Yes | QMessageBox::No, QMessageBox::No ) == QMessageBox::Yes )
+      {
+        connectionParentItem->vacuumGeoPackageDbAction();
+      }
+    }
+    else
+    {
+      QMessageBox::information( nullptr, tr( "Delete Layer" ), tr( "The Layer <b>%1</b> has been deleted successfully." ).arg( mName ) );
```
Comment (Gustry) [inline src/providers/ogr/qgsgeopackagedataitems.cpp]: Wrong capitalization. The **l**ayer

Comment (Gustry) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: Thanks @m-kuhn I forgot this feature in github

File: src/providers/ogr/qgsgeopackagedataitems.cpp [hunk 3]
```diff
@@ -525,9 +525,23 @@ bool QgsGeoPackageAbstractLayerItem::deleteLayer()
   }
   else
   {
-    QMessageBox::information( nullptr, tr( "Delete Layer" ), tr( "Layer <b>%1</b> deleted successfully." ).arg( mName ) );
+
+    QgsGeoPackageConnectionItem *connectionParentItem = qobject_cast<QgsGeoPackageConnectionItem *>( mParent );
+    if ( connectionParentItem )
+    {
+      if ( QMessageBox::question( nullptr, QObject::tr( "Delete Layer" ), QObject::tr( "The layer <b>%1</b> has been deleted successfully."
```
Comment (nyalldawson) [inline src/providers/ogr/qgsgeopackagedataitems.cpp]: Oops, wrong choice. The correct grammar is "was deleted", not "has been" ;)

Comment (m-kuhn) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: Are you sure?
The decision was based on this:
https://github.com/qgis/QGIS/pull/8810#discussion_r245973658

Comment (signedav) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: I have been / was quite sure about that as well. I guess @nyalldawson has to give a lesson about it next Hackfest ;-)

Comment (nyalldawson) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: Yep, I'm sure. I can't point to the rules explaining why sorry... But it just doesn't "sit right".

Comment (m-kuhn) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: Fair enough, @signedav can you prepare a PR?

Comment (signedav) [reply src/providers/ogr/qgsgeopackagedataitems.cpp]: Yes.

**Pull request merged**